cache/class-vary-cache.php

```diff
@@ -370,17 +445,104 @@ public static function set_cookie_expiry( int $expiry ) {
 		self::$cookie_expiry = $expiry;
 	}
 
+	/**
+	 * Sends headers (if needed).
+	 *
+	 * @since   1.0.0
+	 * @access  public
+	 */
 	public static function send_headers() {
-		self::$did_send_headers = true;
+		if ( ! self::$did_send_headers ) {
+			self::$did_send_headers = true;
+
+			$sent_vary = self::send_vary_headers();
+			$sent_cookie = self::set_cookies();
+
+			if ( $sent_vary || $sent_cookie ) {
+				/**
+				 * Vary or Set-Cookie header were sent.
+				 *
+				 * Can be used to handle things like early redirects after a user action and group assignment.
+				 *
+				 * @since 1.0.0
+				 *
+				 * @param boolean $sent_vary Was a Vary header sent?
+				 * @param boolean $sent_cookie Was a Set-Cookie header sent?
+				 */
+				do_action( 'vip_vary_cache_did_send_headers', $sent_vary, $sent_cookie );
+			}
+		}
+	}
+
+	/**
+	 * Determines if cookies need to be set and then sets them.
+	 *
+	 * @since   1.0.0
+	 * @access  public
+	 *
+	 * @return boolean Was at least one cookie set?
+	 */
+	private static function set_cookies() {
+		$sent_cookie = false;
+
+		if ( self::$should_update_group_cookie ) {
+			$sent_cookie = true;
+
+			self::set_group_cookie();
+		}
+
+		if ( self::$should_update_nocache_cookie ) {
+			$sent_cookie = true;
+
+			self::set_nocache_cookie();
+		}
+
+		return $sent_cookie;
+	}
+
+	/**
+	 * Sets the group/segment cookie based on the user's current groupings.
+	 *
+	 * @since   1.0.0
+	 * @access  private
+	 */
+	private static function set_group_cookie() {
```

Comment:
I feel like were having a lot of methods for the various code paths here. It's better for testing but IMO makes it a bit harder to follow (initially). Probably fine as is, but we add any more logic to it might be worth splitting into a separate class (as we had discussed)

Reply:
I actually broke it down to make the flow a little bit easier, but maybe I failed :)

Definitely would be better extracting out in the future.

Reply:
Nah, I think it just varies by person and this case  is totally fine. I just find that sometimes splitting things into too many functions can be a bit hard to trace through on initial read (but might be the better structure once you are familiar/working with it)